File: samples/typescript/resilience/circuit_breaker.ts
type CBState = "CLOSED" | "OPEN" | "HALF_OPEN";
class CircuitBreaker {
  private failures = 0;
  private state: CBState = "CLOSED";
  private openedAt = 0;
  constructor(private threshold = 3, private timeout = 100) {}
  async call<T>(fn: () => Promise<T>): Promise<T> {
    if (this.state === "OPEN") {
      if (Date.now() - this.openedAt >= this.timeout) { this.state = "HALF_OPEN"; }
      else throw new Error(`[CB] OPEN — blocking call`);
    }
    try {
      const r = await fn();
      this.failures = 0; this.state = "CLOSED";
      return r;
    } catch (e) {
      this.failures++;
      if (this.failures >= this.threshold) { this.state = "OPEN"; this.openedAt = Date.now(); }
      throw e;
    }
  }
  getState() { return this.state; }
}
let callCount = 0;
async function flaky(): Promise<string> {
  callCount++;
  if (callCount <= 3) throw new Error(`Downstream fail #${callCount}`);
  return `[LLM] answer on call #${callCount}`;
}
async function main() {
  const cb = new CircuitBreaker(3, 100);
  for (let i = 1; i <= 7; i++) {
    try {
      const r = await cb.call(flaky);
      console.log(`Attempt ${i} [${cb.getState()}]: ${r}`);
    } catch (e: any) {
      console.log(`Attempt ${i} [${cb.getState()}]: ${e.message}`);
    }
    if (i === 4) await new Promise(r => setTimeout(r, 150));
  }
}
main().catch(console.error);

File: samples/python/resilience/circuit_breaker.py
"""
Circuit Breaker Pattern
CLOSED/OPEN/HALF_OPEN states protecting calls to a downstream agent.
"""
import time
from typing import Literal
from langchain_anthropic import ChatAnthropic

llm = ChatAnthropic(model="claude-haiku-4-5-20251001")

State = Literal["CLOSED", "OPEN", "HALF_OPEN"]


class CircuitBreaker:
    def __init__(self, failure_threshold: int = 3, recovery_timeout: float = 2.0):
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.failure_count = 0
        self.state: State = "CLOSED"
        self.opened_at: float = 0.0

    def call(self, fn, *args, **kwargs):
        if self.state == "OPEN":
            elapsed = time.time() - self.opened_at
            if elapsed >= self.recovery_timeout:
                print(f"[CircuitBreaker] Timeout elapsed ({elapsed:.1f}s). Transitioning to HALF_OPEN.")
                self.state = "HALF_OPEN"
            else:
                raise RuntimeError(f"[CircuitBreaker] Circuit is OPEN. Blocking call. ({elapsed:.1f}s elapsed)")

        try:
            result = fn(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure()
            raise e

    def _on_success(self):
        if self.state == "HALF_OPEN":
            print("[CircuitBreaker] Probe succeeded. Transitioning to CLOSED.")
        self.failure_count = 0
        self.state = "CLOSED"

    def _on_failure(self):
        self.failure_count += 1
        print(f"[CircuitBreaker] Failure #{self.failure_count} recorded.")
        if self.failure_count >= self.failure_threshold:
            self.state = "OPEN"
            self.opened_at = time.time()
            print(f"[CircuitBreaker] Threshold reached. Transitioning to OPEN.")


call_count = 0


def flaky_downstream(prompt: str) -> str:
    global call_count
    call_count += 1
    if call_count <= 3:
        raise ConnectionError(f"Downstream unavailable (call #{call_count})")
    response = llm.invoke(prompt)
    return response.content if hasattr(response, "content") else str(response)


def main():
    cb = CircuitBreaker(failure_threshold=3, recovery_timeout=2.0)
    prompt = "In one sentence, what is a circuit breaker in software?"

    for attempt in range(1, 8):
        print(f"\n--- Attempt {attempt} | CB state: {cb.state} ---")
        try:
            result = cb.call(flaky_downstream, prompt)
            print(f"Success: {result}")
        except RuntimeError as e:
            print(f"Blocked: {e}")
        except ConnectionError as e:
            print(f"Downstream error: {e}")

        if attempt == 4:
            print(f"\n[main] Waiting {cb.recovery_timeout}s for circuit recovery timeout...")
            time.sleep(cb.recovery_timeout + 0.1)


if __name__ == "__main__":
    main()
